Updates to DRA Partitionable Devices feature

mortent opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR updates the Partitionable Devices feature according to the changes described in https://github.com/kubernetes/enhancements/pull/5515

#### Which issue(s) this PR is related to:

Related-to: https://github.com/kubernetes/enhancements/issues/4815

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Updates to the Partitionable Devices feature which allows for referencing counter sets across different ResourceSlices within the same resource pool.

Devices from incomplete pools are no longer considered for allocation.

This contains backwards incompatible changes to the Partitionable Devices alpha feature, so any ResourceSlices that uses the feature should be removed prior to upgrading or downgrading between 1.34 and 1.35.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4815-dra-partitionable-devices

mortent committed:
DRA: Update types for Partitionable Devices

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -251,10 +251,6 @@ const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a
 
-// Defines the max number of shared counters that can be specified
-// in a ResourceSlice. The number is summed up across all sets.
-const ResourceSliceMaxSharedCounters = 32
```
I'd expect this in the first commit.

pohly commented on test/integration/dra/dra_test.go:
```diff
@@ -1246,24 +1246,38 @@ func testResourceClaimDeviceStatus(tCtx ktesting.TContext, enabled bool) {
+		"resourceslice-with-shared-counters": {
+			resourceSliceGenFunc: NewMaxResourceSliceWithSharedCounters,
+		},
+	} {
```
I arrived at a similar solution in the device taints PR, except that I turned NewMaxResourceSlice into `func NewMaxResourceSlices() map[string]*resourceapi.ResourceSlice`, i.e. name+slice as result. IMHO a bit simpler:

https://github.com/pohly/kubernetes/blob/6daab8c9e7c205f2a5e06c6e88f5ca9496e238ca/test/integration/dra/dra_test.go#L1257-L1262

pohly reviewed (changes requested):
First pass, without looking into some of the implementation aspects yet (in particular the allocator).

liggitt commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -171,19 +174,22 @@ type ResourceSliceSpec struct {
 	// +optional
 	// +listType=atomic
 	// +featureGate=DRAPartitionableDevices
+	// +zeroOrOneOf=ResourceSliceType
```
same question for the fields inside CounterSet (name format, counters max entries, etc)

liggitt commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -199,7 +205,7 @@ type CounterSet struct {
+	// The maximum number of counters is 32.
 	//
 	// +required
 	// +k8s:required
```
(for follow-up, once https://github.com/jpbetz/validation-gen/issues/208 is done we can add `// +k8s:maxProperties=32` here and in `DeviceCounterConsumption`)

liggitt commented on staging/src/k8s.io/api/resource/v1/types.go:
once we get v1/types.go in shape w.r.t. any declarative validation updates, mirror those to other versioned types.go files

liggitt commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/validation.go:
```diff
@@ -0,0 +1,82 @@
+	for _, slice := range pool.Slices {
+		if slice.SharedCounters != nil && slice.Devices != nil {
+			return fmt.Errorf("found slice with both counter sets and devices in pool %q", name)
+		}
```
+1 for dropping this check from here, the controller doesn't actually care about that

liggitt commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/validation.go:
```diff
@@ -0,0 +1,82 @@
+			for _, dcc := range device.ConsumesCounters {
+				counterSet, found := counterSets[dcc.CounterSet]
+				if !found {
+					return fmt.Errorf("counter set %q referenced by device %q not found", dcc.CounterSet, device.Name)
```
do we care about including the name of the resourceslice where we found the invalid thing?

aaron-prindle commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -289,10 +313,9 @@ type Device struct {
+	// devices in a ResourceSlice is 2048.
 	//
 	// +optional
 	// +listType=atomic
```
Yes the tags above would work for the validation of "max items 2", "items in map unique" - with the caveat being that you would also need to use the `+k8s:listType=atomic` DV tag as logically you need both the DV `+k8s:listType=...` and `+k8s:listMapKey=...` for validation gen.  Eg:

```
// +k8s:maxItems=2
// +k8s:unique=map
// +k8s:listMap=atomic
// +k8s:listMapKey=counterSet
```

DV does not have a tag for this validation though:

> The maximum total number of consumed counters across
	// devices in a ResourceSlice is 2048.

aaron-prindle commented on staging/src/k8s.io/api/resource/v1beta1/types.go:
```diff
@@ -171,19 +174,22 @@ type ResourceSliceSpec struct {
 	// +listType=atomic
 	// +featureGate=DRAPartitionableDevices
+	// +zeroOrOneOf=ResourceSliceType
 	SharedCounters []CounterSet `json:"sharedCounters,omitempty" protobuf:"bytes,8,name=sharedCounters"`
```
For enforcing "maximum number of counter sets is 8", can use the DV tag:
// +k8s:maxItems=8

aaron-prindle commented:
Also want to note that for any DV tag added to a field, that field should also have either a +k8s:optional or +k8s:required tag as well (this would be in addition to the +optional or +required tag on the field).

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/validation.go:
```diff
@@ -0,0 +1,82 @@
+				}
+				for counterName := range dcc.Counters {
+					if _, found := counterSet.Counters[counterName]; !found {
+						return fmt.Errorf("counter %q referenced by device %q not found in counter set %q", counterName, device.Name, counterSet.Name)
```
Same as above

mortent commented on pkg/apis/resource/types.go:
```diff
@@ -251,6 +255,28 @@ const BindingFailureConditionsMaxSize = 4
+// Defines the maximum number of counters that can be defined
+// in device counter consumptions across all devices in a
+// ResourceSlice.
+const ResourceSliceMaxConsumedCountersPerResourceSlice = 2048
```
Yes, we should do the same here. With a 64 device limit, the current thresholds results in a worst case of 651,357 bytes, and 809,053 bytes without the resourceslice-wide limit. I will update the PR.

liggitt replied:
will the calculations being done for max size with 64 devices still be ok if the devices contain their max taints AND max consumed counters?

liggitt commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -703,35 +703,23 @@ func validateResourceSliceSpec(spec, oldSpec *resource.ResourceSliceSpec, fldPat
+
+	maxDevices := resource.ResourceSliceMaxDevices
+	if haveConsumesCounters(spec) {
+		maxDevices = resource.ResourceSliceMaxDevicesWithConsumesCounters
```
same comment as at https://github.com/kubernetes/kubernetes/pull/134152#discussion_r2486630747

once this PR and that PR need to both live in master, should we combine their device count limit detection into a single function with a generic constant / name like `ResourceSliceMaxDevicesWithComplexFields` or something?

johnbelamaric replied:
+1

pohly replied:
I agree that we can combine both features with a limit of 64, assuming that total device size doesn't get too big - to be verified once the first feature is merged and while adapting the second.

mortent replied:
Yeah, I've updated this to follow the same pattern as in https://github.com/kubernetes/kubernetes/pull/134152 for now. But it needs to be updated if https://github.com/kubernetes/kubernetes/pull/134152 merges first (which I think it probably will).

I don't think the size will be a problem, but we definitely need to verify.

liggitt reviewed (commented):
API bits look good to me.

One non-blocking comment about coordination with https://github.com/kubernetes/kubernetes/pull/134152#discussion_r2486630747

I'd like a DV reviewer to do a final pass on the added tags, especially with an eye towards whether we're consistent with the same tags on the same fields across versions (hopefully we have a linter or test that would catch if we weren't?)

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/validation.go:
```diff
@@ -0,0 +1,82 @@
+		}
+		for _, counterSet := range slice.SharedCounters {
+			if _, found := counterSets[counterSet.Name]; found {
+				return fmt.Errorf("found duplicate counter set %q in pool %q", counterSet.Name, name)
```
We could include the slice index because that is something that might help the DRA driver author track down the error, but it's not important. Let's leave it out.

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/experimental/pools_experimental.go:
```diff
@@ -118,33 +124,61 @@ func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node
+			result = append(result, &Pool{
+				PoolID:       poolID,
+				IsIncomplete: true,
+			})
```
So for AllocationMode: All we error out if the pool is incomplete: https://github.com/kubernetes/dynamic-resource-allocation/blob/42e21234a6d933c0a8bc0ee16a8a214463d45646/structured/internal/experimental/allocator_experimental.go#L535-L537

So if we remove it, then allocationMode: All might succeed without actually considering all devices. So I don't think we want to do that.

But ref the next discussion. Since we return an error here, it seems like a single incomplete pool for a single node would block scheduling of the pod on any nodes in the cluster, even though the incomplete pool might not be relevant. Seems like ideally we want similar behavior here as we are discussing for invalid pools. Although we can probably assume in most cases that incomplete pools become complete relatively quickly.

pohly replied:
> So if we remove it, then allocationMode: All might succeed without actually considering all devices. So I don't think we want to do that.

Ack. Worth adding a source comment.

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/experimental/allocator_experimental.go:
```diff
@@ -1411,83 +1421,87 @@ func taintTolerated(taint resourceapi.DeviceTaint, request requestAccessor) bool
 	// calculated.
 	alloc.mutex.RLock()
-	availableCountersForSlice, found := alloc.availableCounters[sliceName]
+	availableCountersForPool, found := alloc.availableCounters[poolName]
```
Yeah, I did try this as part of a larger change, but ran into issues where it created issues with logging since the encoding/json package doesn't handle non-string keys in maps. But that shouldn't be an issue here, so I have updated the PR to use UniqueString.

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/api/types.go:
```diff
@@ -78,3 +80,5 @@ type DeviceCounterConsumption struct {
 	Counters   map[string]resourceapi.Counter `json:",omitempty"`
 }
+
+var ErrInvalidResourcePoolEncountered = fmt.Errorf("invalid resource pools were encountered")
```
Thanks, this makes sense. I have updated the PR.

mortent commented:
> These should be added for the v1beta1, v1beta2, and v1 types. **NOTE: REMOVE the tags from the internal types.go (pkg/apis/resource/types.go) as DV tags only go on the versioned types.**
> 
> To fully plumb these tags, you will need to add tests in declarative_validation_test.go (I believe you had the tests there prior so add them back) and mark the analogous hand-written validation.go errors as MarkedCoveredByDeclarative.

@aaron-prindle I have updated types and added back the relevant declarative validation tests. Could you do a review of that part of the PR?

aaron-prindle commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -171,19 +174,26 @@ type ResourceSliceSpec struct {
+	// +k8s:listMapKey=name
 	// +featureGate=DRAPartitionableDevices
+	// +zeroOrOneOf=ResourceSliceType
+	// +k8s:maxItems=8
```
All DV fields need to be marked +k8s:optional or +k8s:required.  This also needs a:
```
// +k8s:optional
```
tag.  This should be done across all API versions

mortent commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -909,14 +872,15 @@ func validateDeviceCounterConsumption(deviceCounterConsumption resource.DeviceCo
 	}
 	if len(deviceCounterConsumption.Counters) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("counters"), ""))
+		allErrs = append(allErrs, field.Required(fldPath.Child("counters"), "").MarkCoveredByDeclarative())
```
Fixed.

mortent commented on pkg/apis/resource/validation/validation_resourceslice_test.go:
```diff
@@ -756,80 +828,129 @@ func TestValidateResourceSlice(t *testing.T) {
+		"missing-counter-consumes-counter": {
 			wantFailures: field.ErrorList{
-				field.Invalid(field.NewPath("spec", "devices").Index(0).Child("consumesCounters").Index(0).Child("counterSet"), "fake", "must reference a counterSet defined in the ResourceSlice sharedCounters"),
+				field.Required(field.NewPath("spec", "devices").Index(0).Child("consumesCounters").Index(0).Child("counters"), "").MarkCoveredByDeclarative(),
```
Fixed

mortent commented:
/test pull-kubernetes-unit

pohly reviewed (approved):
DRA parts look good to me, doc comment needs one minor update.

/approve

mortent committed:
Run make update

mortent committed:
DRA: Update validation for Partitionable Devices

mortent committed:
DRA: Update ResourceSlice max size tests for Partitionable Devices

mortent commented on pkg/apis/resource/types.go:
```diff
@@ -243,14 +249,27 @@ type ResourcePool struct {
 const ResourceSliceMaxSharedCapacity = 128
 const ResourceSliceMaxDevices = 128
-const ResourceSliceMaxDevicesWithTaints = 64
+const ResourceSliceMaxDevicesWithTaintsOrConsumesCounters = 64
```
Fixed.

mortent committed:
DRA: Update allocator for Partitionable Devices

mortent commented:
@pohly @liggitt There was a conflict, so I had to rebase. Can one of you re-add the lgtm?

mortent committed:
DRA: Add validation in resourceslicecontroller for Partitionable Devices

mortent committed:
DRA: Update integration tests for Partitionable Devices

mortent committed:
DRA: Update e2e tests for Partitionable Devices

liggitt commented:
/lgtm
